File: DataManager/Services/HealthDataService.cs
using Grpc.Core;
using Google.Protobuf.WellKnownTypes;
using DataManager.Data;
using DataManager.Models;
using DataManager.Grpc;
using Microsoft.AspNetCore.Builder.Extensions;
using System.Globalization;
using DataManager.MQTT;
using System.Text.Json;

namespace DataManager.Services;

public class HealthDataService : HealthData.HealthDataBase
{
    private readonly HealthRecordDbContext _context;
    private readonly MqttPublisher _mqttPublisher;
    private readonly ILogger<HealthDataService> _logger;

    public HealthDataService(
        HealthRecordDbContext context,
        ILogger<HealthDataService> logger,
        MqttPublisher mqttPublisher
    ) {
        _logger = logger;
        _context = context;
        _mqttPublisher = mqttPublisher;
    }

    public override async Task<Empty> CreateHealthRecord(
        HealthRecordRequest request,
        ServerCallContext context
    ){
        try
        {
            int recordId = await _context.CreateHealthRecordAsync(new HealthRecordModel
            {
                athleteId = request.AthleteId,
                timestamp = request.Timestamp.ToDateTime(),
                heartRate = request.HeartRate,
                bodyTemperature = request.BodyTemperature,
                bloodPressure = request.BloodPressure,
                bloodOxygen = request.BloodOxygen,
                stepCount = request.StepCount,
                activityStatus = (ActivityStatusModel)request.ActivityStatus,
                latitude = request.Latitude,
                longitude = request.Longitude,
                secureTransmissionStatus = request.SecureTransmissionStatus
            });

            HealthRecordModel record = new HealthRecordModel
            {
                recordId = recordId,
                athleteId = request.AthleteId,
                timestamp = request.Timestamp.ToDateTime(),
                heartRate = request.HeartRate,
                bodyTemperature = request.BodyTemperature,
                bloodPressure = request.BloodPressure,
                bloodOxygen = request.BloodOxygen,
                stepCount = request.StepCount,
                activityStatus = (ActivityStatusModel)request.ActivityStatus,
                latitude = request.Latitude,
                longitude = request.Longitude,
                secureTransmissionStatus = request.SecureTransmissionStatus
            };

            await _mqttPublisher.PublishAsync("health/records", record, retain: true);

            _logger.LogInformation($"Created health record with ID: {recordId}");
            return new Empty();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error while creating health record");
            throw new RpcException(new Status(
                StatusCode.Internal,
                $"Failed to create health record: {ex.Message}"
            ));
        }
    }


    public override async Task<HealthRecord> GetHealthRecord(HealthRecordIdRequest request, ServerCallContext context)
    {
        var record = await _context.GetHealthRecordAsync(request.RecordId);
        if (record == null)
        {
            throw new RpcException(new Status(StatusCode.NotFound, $"Record with ID {request.RecordId} not found."));
        }
        return new HealthRecord
            {
                RecordId = record.recordId,
                AthleteId = record.athleteId,
                Timestamp = Timestamp.FromDateTime(record.timestamp.ToUniversalTime()),
                HeartRate = record.heartRate,
                BodyTemperature = record.bodyTemperature,
                BloodPressure = record.bloodPressure,
                BloodOxygen = record.bloodOxygen,
                StepCount = record.stepCount,
                ActivityStatus = (ActivityStatus)record.activityStatus,
                Latitude = record.latitude,
                Longitude = record.longitude,
                SecureTransmissionStatus = record.secureTransmissionStatus
            };
    }

    public override async Task<HealthRecordResponse> GetAllHealthRecords(Empty request, ServerCallContext context)
    {
        var records = await _context.GetAllHealthRecordsAsync();

        var response = new HealthRecordResponse();
        if (records != null)
        {
            foreach (var record in records)
            {
                response.Records.Add(new HealthRecord
                {
                    RecordId = record.recordId,
                    AthleteId = record.athleteId,
                    Timestamp = Timestamp.FromDateTime(record.timestamp.ToUniversalTime()),
                    HeartRate = record.heartRate,
                    BodyTemperature = record.bodyTemperature,
                    BloodPressure = record.bloodPressure,
                    BloodOxygen = record.bloodOxygen,
                    StepCount = record.stepCount,
                    ActivityStatus = (ActivityStatus)record.activityStatus,
                    Latitude = record.latitude,
                    Longitude = record.longitude,
                    SecureTransmissionStatus = record.secureTransmissionStatus
                });
            }
        }

        return await Task.FromResult(response);
    }

    public override async Task<Message> UpdateHealthRecord(HealthRecord request, ServerCallContext context)
    {
        bool update = await _context.UpdateHealthRecordAsync(new HealthRecordModel
        {
            recordId = request.RecordId,
            athleteId = request.AthleteId,
            timestamp = request.Timestamp.ToDateTime(),
            heartRate = request.HeartRate,
            bodyTemperature = request.BodyTemperature,
            bloodPressure = request.BloodPressure,
            bloodOxygen = request.BloodOxygen,
            stepCount = request.StepCount,
            activityStatus = (ActivityStatusModel)request.ActivityStatus,
            latitude = request.Latitude,
            longitude = request.Longitude,
            secureTransmissionStatus = request.SecureTransmissionStatus
        });

        if (update == false)
        {
            throw new RpcException(new Status(StatusCode.NotFound, $"Record with ID {request.RecordId} not found."));
        }

        return await Task.FromResult(new Message { RecordId = request.RecordId, Message_ = "Record was updated" });
    }

    public override async Task<Message> DeleteHealthRecord(HealthRecordIdRequest request, ServerCallContext context)
    {
        var delete = await _context.DeleteHealthRecordAsync(request.RecordId);
        if (delete == false)
        {
            throw new RpcException(new Status(StatusCode.NotFound, $"Record with ID {request.RecordId} not found."));
        }

        return await Task.FromResult(new Message { RecordId = request.RecordId, Message_ = "Record was deleted" });
    }

    public override async Task<HealthRecordResponse> GetAthleteHealthRecords(AthleteFilter request, ServerCallContext context)
    {
        var records = await _context.GetAthleteHealthRecordsAsync(request.AthleteId, (ActivityStatusModel)request.ActivityStatus, request.StartTime, request.EndTime);
        
        var response = new HealthRecordResponse();
        if (records != null)
        {
            foreach (var record in records)
            {
                response.Records.Add(new HealthRecord
                {
                    RecordId = record.recordId,
                    AthleteId = record.athleteId,
                    Timestamp = Timestamp.FromDateTime(record.timestamp.ToUniversalTime()),
                    HeartRate = record.heartRate,
                    BodyTemperature = record.bodyTemperature,
                    BloodPressure = record.bloodPressure,
                    BloodOxygen = record.bloodOxygen,
                    StepCount = record.stepCount,
                    ActivityStatus = (ActivityStatus)record.activityStatus,
                    Latitude = record.latitude,
                    Longitude = record.longitude,
                    SecureTransmissionStatus = record.secureTransmissionStatus
                });
            }
        }

        _logger.LogInformation($"request");

        return await Task.FromResult(response);
    }

}

File: gateway/src/services/healthDataService.ts
import { Timestamp } from '../grpc/google/protobuf/timestamp';
import { client } from '../grpc/healtDataClient';
import { AthleteFilter, HealthRecord, HealthRecordIdRequest, HealthRecordRequest, Message } from '../grpc/healthData';
import { HealthRecordAggregation } from '../interface/healthDataInterface';

async function CreateHealthRecord(r: HealthRecordRequest): Promise<void> {
  const record = {
    ...r,
    timestamp: r.timestamp ? toProtoTimestamp(new Date(r.timestamp)) : undefined,
  };
  return new Promise((resolve, reject) => {
    client.CreateHealthRecord(record, (error: any, _response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

async function GetHealthRecords(recordId: number): Promise<HealthRecord> {
  const request: HealthRecordIdRequest = { recordId };
  return new Promise((resolve, reject) => {
    client.GetHealthRecord(request, (error: any, response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve(response as HealthRecord);
      }
    });
  });
}

async function GetAllHealthRecords(): Promise<HealthRecord[]> {
  return new Promise((resolve, reject) => {
    client.GetAllHealthRecords({}, (error: any, response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve(response.records as HealthRecord[]);
      }
    });
  });
}

async function UpdateHealthRecord(r: HealthRecord): Promise<Message> {
  const record = {
    ...r,
    timestamp: r.timestamp ? toProtoTimestamp(new Date(r.timestamp)) : undefined,
  };
  return new Promise((resolve, reject) => {
    client.UpdateHealthRecord(record, (error: any, response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve(response as Message);
      }
    });
  });
}

async function DeleteHealthRecord(recordId: number): Promise<Message> {
  const request: HealthRecordIdRequest = { recordId };
  return new Promise((resolve, reject) => {
    client.DeleteHealthRecord(request, (error: any, response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve(response as Message);
      }
    });
  });
}

async function GetAthleteHealthRecords(req: AthleteFilter): Promise<HealthRecord[]> {
  if (!req.athleteId || req.athleteId == "") {
    throw new Error("athleteId is required");
  }
  const filter = {
    athleteId: req.athleteId,
    activityStatus: req.activityStatus,
    startTime: req.startTime ? toProtoTimestamp(new Date(req.startTime)) : undefined,
    endTime: req.endTime ? toProtoTimestamp(new Date(req.endTime)) : undefined,
  };
  return new Promise((resolve, reject) => {
    console.log("Filter:", filter);
    client.GetAthleteHealthRecords(filter, (error: any, response: any) => {
      if (error) {
        reject(error);
      } else {
        resolve(response.records as HealthRecord[]);
      }
    });
  });
}

async function AggregateAthleteRecords(
  filter: AthleteFilter
): Promise<HealthRecordAggregation | null> {
  const records = await GetAthleteHealthRecords(filter);
  if (!records || records.length === 0) {
    return null;
  }

  const systolics: number[] = [];
  const diastolics: number[] = [];

  const heartRateValues = records.map((r) => r.heartRate);
  const tempValues = records.map((r) => r.bodyTemperature);
  const oxygenValues = records.map((r) => r.bloodOxygen);
  const stepValues = records.map((r) => r.stepCount);

  const recordSummaries = records.map((r) => ({
    timestamp: r.timestamp!,
    bloodPressure: r.bloodPressure,
    latitude: r.latitude,
    longitude: r.longitude,
    secureTransmissionStatus: r.secureTransmissionStatus,
  }));

  records.forEach((r) => {
    const parsed = parseBloodPressure(r.bloodPressure);
    if (parsed) {
      systolics.push(parsed.systolic);
      diastolics.push(parsed.diastolic);
    }
  });

  const bpStats = {
    systolic: {
      min: Math.min(...systolics),
      max: Math.max(...systolics),
    },
    diastolic: {
      min: Math.min(...diastolics),
      max: Math.max(...diastolics),
    },
  };

  return {
    athleteId: filter.athleteId,
    activityStatus: filter.activityStatus,
    startTime: filter.startTime,
    endTime: filter.endTime,

    heartRate: calcStats(heartRateValues),
    bodyTemperature: calcStats(tempValues),
    bloodOxygen: calcStats(oxygenValues),
    stepCount: {
      sum: stepValues.reduce((a, b) => a + b, 0),
    },
    bloodPressure: {
      min: `${bpStats.systolic.min}/${bpStats.diastolic.min}`,
      max: `${bpStats.systolic.max}/${bpStats.diastolic.max}`,
    },

    records: recordSummaries,
  };
}

function calcStats(values: number[]) {
  return {
    min: Math.min(...values),
    max: Math.max(...values),
    avg: values.reduce((a, b) => a + b, 0) / values.length,
  };
}

function toProtoTimestamp(date: Date): Timestamp | null {

  if (!date) {
    return null;
  }
  return {
    seconds: Math.floor(date.getTime() / 1000),
    nanos: (date.getTime() % 1000) * 1_000_000,
  };
}

function parseBloodPressure(bp: string): { systolic: number; diastolic: number } | null {
  if (!bp) return null;
  const parts = bp.split("/");
  if (parts.length !== 2) return null;
  const systolic = parseInt(parts[0], 10);
  const diastolic = parseInt(parts[1], 10);
  if (isNaN(systolic) || isNaN(diastolic)) return null;
  return { systolic, diastolic };
}


export { AggregateAthleteRecords, CreateHealthRecord, DeleteHealthRecord, GetAllHealthRecords, GetAthleteHealthRecords, GetHealthRecords, UpdateHealthRecord };
